vcstools/gfit.py

```diff
@@ -431,27 +458,43 @@ def chsq(observed_values, expected_values, err):
         return [fit, redchisq, best_bic, popt, pcov, comp_dict, comp_idx]
 
     # SigmaClip isn't perfect. Use these next function to fix bad clips
-    def _fill_clipped_prof(self, search_scope=None, nan_type=0.):
+    def _fill_off_pulse_prof(self, search_scope=None):
         """
         Intended for use on noisy profiles. Fills nan values that are surrounded by non-nans to avoid discontinuities
         in the profile
         """
-        length = len(self._clipped_prof)
+        length = len(self._off_pulse_prof)
         if search_scope is None:
             # Search 5% ahead for non-nans
-            search_scope = round(length*0.05)
-        search_scope = np.linspace(1, search_scope, search_scope, dtype=int)
+            search_scope = round(length*0.2)
+        search_scope = np.linspace(search_scope, 1, search_scope, dtype=int)
+        # Search scope = [x, x-1, x-2, ..., 1]
 
         # Loop over all values in clipped profile
-        for i, val in enumerate(self._clipped_prof):
-            if val == nan_type and not (i+max(search_scope)) >= length:
-                #look 'search_scope' indices ahead for non-nans
-                for j in sorted(search_scope, reverse=True):
-                    #fill in nans
-                    if self._clipped_prof[i+j]==nan_type:
-                        for k in range(j):
-                            self._clipped_prof[i+k]=nan_type
+        for i, val in enumerate(self._off_pulse_prof):
+            # If current index is on-pulse
+            if np.isnan(val):
+                fill_i=0
+                # Loop over search scope
+                for search_i in search_scope: 
+                    # Make a copy of the profile and roll back 
+                    rolled_prof = np.roll(self._off_pulse_prof, -search_i)
+                    # If we find more on-pulse within the search scope and break
+                    if np.isnan(rolled_prof[i]):
+                        fill_i = search_i
                         break
+                # Fill in the on-pulse
+                for j in range(fill_i):
+                    fill_idx = i+j if i+j<len(self._off_pulse_prof) else i+j-len(self._off_pulse_prof)
+                    self._off_pulse_prof[fill_idx] = np.nan
```

Comment:
This also confuses me but after a few reads, I think I get it. It finds all the on pulse then trys to find more the on pulse within 20% and fills in all values between it.